File: binding/src/libqcolorspace.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qcolorspace.html
pub const qcolorspace = struct {

    /// New constructs a new QColorSpace object.
    pub fn New() QtC.QColorSpace {
        return qtc.QColorSpace_new();
    }


    /// New2 constructs a new QColorSpace object.
    pub fn New2(namedColorSpace: i32) QtC.QColorSpace {
        return qtc.QColorSpace_new2(@intCast(namedColorSpace));
    }


    /// New3 constructs a new QColorSpace object.
    pub fn New3(whitePoint: QtC.QPointF, transferFunction: i32) QtC.QColorSpace {
        return qtc.QColorSpace_new3(@ptrCast(whitePoint), @intCast(transferFunction));
    }


    /// New4 constructs a new QColorSpace object.
    pub fn New4(whitePoint: QtC.QPointF, transferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_new4(@ptrCast(whitePoint), @ptrCast(transferFunctionTable));
    }


    /// New5 constructs a new QColorSpace object.
    pub fn New5(primaries: i32, transferFunction: i32) QtC.QColorSpace {
        return qtc.QColorSpace_new5(@intCast(primaries), @intCast(transferFunction));
    }


    /// New6 constructs a new QColorSpace object.
    pub fn New6(primaries: i32, gamma: f32) QtC.QColorSpace {
        return qtc.QColorSpace_new6(@intCast(primaries), gamma);
    }


    /// New7 constructs a new QColorSpace object.
    pub fn New7(primaries: i32, transferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_new7(@intCast(primaries), @ptrCast(transferFunctionTable));
    }


    /// New8 constructs a new QColorSpace object.
    pub fn New8(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunction: i32) QtC.QColorSpace {
        return qtc.QColorSpace_new8(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @intCast(transferFunction));
    }


    /// New9 constructs a new QColorSpace object.
    pub fn New9(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_new9(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @ptrCast(transferFunctionTable));
    }


    /// New10 constructs a new QColorSpace object.
    pub fn New10(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, redTransferFunctionTable: []const u8, greenTransferFunctionTable: []const u8, blueTransferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_new10(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @ptrCast(redTransferFunctionTable), @ptrCast(greenTransferFunctionTable), @ptrCast(blueTransferFunctionTable));
    }


    /// New11 constructs a new QColorSpace object.
    pub fn New11(colorSpace: ?*anyopaque) QtC.QColorSpace {
        return qtc.QColorSpace_new11(@ptrCast(colorSpace));
    }


    /// New12 constructs a new QColorSpace object.
    pub fn New12(whitePoint: QtC.QPointF, transferFunction: i32, gamma: f32) QtC.QColorSpace {
        return qtc.QColorSpace_new12(@ptrCast(whitePoint), @intCast(transferFunction), gamma);
    }


    /// New13 constructs a new QColorSpace object.
    pub fn New13(primaries: i32, transferFunction: i32, gamma: f32) QtC.QColorSpace {
        return qtc.QColorSpace_new13(@intCast(primaries), @intCast(transferFunction), gamma);
    }


    /// New14 constructs a new QColorSpace object.
    pub fn New14(whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque, transferFunction: i32, gamma: f32) QtC.QColorSpace {
        return qtc.QColorSpace_new14(@ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint), @intCast(transferFunction), gamma);
    }


    pub fn OperatorAssign(self: ?*anyopaque, colorSpace: ?*anyopaque) void {
        qtc.QColorSpace_OperatorAssign(@ptrCast(self), @ptrCast(colorSpace));
    }

    pub fn Swap(self: ?*anyopaque, colorSpace: ?*anyopaque) void {
        qtc.QColorSpace_Swap(@ptrCast(self), @ptrCast(colorSpace));
    }

    pub fn Primaries(self: ?*anyopaque, ) i32 {
        return qtc.QColorSpace_Primaries(@ptrCast(self));
    }

    pub fn TransferFunction(self: ?*anyopaque, ) i32 {
        return qtc.QColorSpace_TransferFunction(@ptrCast(self));
    }

    pub fn Gamma(self: ?*anyopaque, ) f32 {
        return qtc.QColorSpace_Gamma(@ptrCast(self));
    }

    pub fn Description(self: ?*anyopaque, , allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QColorSpace_Description(@ptrCast(self));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qcolorspace.Description: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetDescription(self: ?*anyopaque, description: []const u8) void {
        const description_str = qtc.libqt_string{
    .len = description.len,
    .data = description.ptr,
};
qtc.QColorSpace_SetDescription(@ptrCast(self), description_str);
    }

    pub fn SetTransferFunction(self: ?*anyopaque, transferFunction: i32) void {
        qtc.QColorSpace_SetTransferFunction(@ptrCast(self), @intCast(transferFunction));
    }

    pub fn SetTransferFunction2(self: ?*anyopaque, transferFunctionTable: []const u8) void {
        qtc.QColorSpace_SetTransferFunction2(@ptrCast(self), @ptrCast(transferFunctionTable));
    }

    pub fn SetTransferFunctions(self: ?*anyopaque, redTransferFunctionTable: []const u8, greenTransferFunctionTable: []const u8, blueTransferFunctionTable: []const u8) void {
        qtc.QColorSpace_SetTransferFunctions(@ptrCast(self), @ptrCast(redTransferFunctionTable), @ptrCast(greenTransferFunctionTable), @ptrCast(blueTransferFunctionTable));
    }

    pub fn WithTransferFunction(self: ?*anyopaque, transferFunction: i32) QtC.QColorSpace {
        return qtc.QColorSpace_WithTransferFunction(@ptrCast(self), @intCast(transferFunction));
    }

    pub fn WithTransferFunction2(self: ?*anyopaque, transferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_WithTransferFunction2(@ptrCast(self), @ptrCast(transferFunctionTable));
    }

    pub fn WithTransferFunctions(self: ?*anyopaque, redTransferFunctionTable: []const u8, greenTransferFunctionTable: []const u8, blueTransferFunctionTable: []const u8) QtC.QColorSpace {
        return qtc.QColorSpace_WithTransferFunctions(@ptrCast(self), @ptrCast(redTransferFunctionTable), @ptrCast(greenTransferFunctionTable), @ptrCast(blueTransferFunctionTable));
    }

    pub fn SetPrimaries(self: ?*anyopaque, primariesId: i32) void {
        qtc.QColorSpace_SetPrimaries(@ptrCast(self), @intCast(primariesId));
    }

    pub fn SetPrimaries2(self: ?*anyopaque, whitePoint: ?*anyopaque, redPoint: ?*anyopaque, greenPoint: ?*anyopaque, bluePoint: ?*anyopaque) void {
        qtc.QColorSpace_SetPrimaries2(@ptrCast(self), @ptrCast(whitePoint), @ptrCast(redPoint), @ptrCast(greenPoint), @ptrCast(bluePoint));
    }

    pub fn SetWhitePoint(self: ?*anyopaque, whitePoint: QtC.QPointF) void {
        qtc.QColorSpace_SetWhitePoint(@ptrCast(self), @ptrCast(whitePoint));
    }

    pub fn WhitePoint(self: ?*anyopaque, ) QtC.QPointF {
        return qtc.QColorSpace_WhitePoint(@ptrCast(self));
    }

    pub fn TransformModel(self: ?*anyopaque, ) u8 {
        return qtc.QColorSpace_TransformModel(@ptrCast(self));
    }

    pub fn ColorModel(self: ?*anyopaque, ) i32 {
        return qtc.QColorSpace_ColorModel(@ptrCast(self));
    }

    pub fn Detach(self: ?*anyopaque, ) void {
        qtc.QColorSpace_Detach(@ptrCast(self));
    }

    pub fn IsValid(self: ?*anyopaque, ) bool {
        return qtc.QColorSpace_IsValid(@ptrCast(self));
    }

    pub fn IsValidTarget(self: ?*anyopaque, ) bool {
        return qtc.QColorSpace_IsValidTarget(@ptrCast(self));
    }

    pub fn FromIccProfile(iccProfile: []u8) QtC.QColorSpace {
        const iccProfile_str = qtc.libqt_string{
    .len = iccProfile.len,
    .data = iccProfile.ptr,
};
return qtc.QColorSpace_FromIccProfile(iccProfile_str);
    }

    pub fn IccProfile(self: ?*anyopaque, , allocator: std.mem.Allocator) []u8 {
        const _bytearray: qtc.libqt_string = qtc.QColorSpace_IccProfile(@ptrCast(self));
defer qtc.libqt_string_free(&_bytearray);
const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("qcolorspace.IccProfile: Memory allocation failed");
@memcpy(_ret, _bytearray.data[0.._bytearray.len]);
return  _ret;
    }

    pub fn TransformationToColorSpace(self: ?*anyopaque, colorspace: ?*anyopaque) QtC.QColorTransform {
        return qtc.QColorSpace_TransformationToColorSpace(@ptrCast(self), @ptrCast(colorspace));
    }

    pub fn OperatorQVariant(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QColorSpace_OperatorQVariant(@ptrCast(self));
    }

    pub fn SetTransferFunction22(self: ?*anyopaque, transferFunction: i32, gamma: f32) void {
        qtc.QColorSpace_SetTransferFunction22(@ptrCast(self), @intCast(transferFunction), gamma);
    }

    pub fn WithTransferFunction22(self: ?*anyopaque, transferFunction: i32, gamma: f32) QtC.QColorSpace {
        return qtc.QColorSpace_WithTransferFunction22(@ptrCast(self), @intCast(transferFunction), gamma);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QColorSpace_Delete(@ptrCast(self));
    }
};

/// https://doc.qt.io/qt-6/colorspace.html#types
pub const enums = struct {
    pub const NamedColorSpace = enum {
        pub const SRgb: i32 = 1;
        pub const SRgbLinear: i32 = 2;
        pub const AdobeRgb: i32 = 3;
        pub const DisplayP3: i32 = 4;
        pub const ProPhotoRgb: i32 = 5;
        pub const Bt2020: i32 = 6;
        pub const Bt2100Pq: i32 = 7;
        pub const Bt2100Hlg: i32 = 8;
    };

    pub const Primaries = enum {
        pub const Custom: i32 = 0;
        pub const SRgb: i32 = 1;
        pub const AdobeRgb: i32 = 2;
        pub const DciP3D65: i32 = 3;
        pub const ProPhotoRgb: i32 = 4;
        pub const Bt2020: i32 = 5;
    };

    pub const TransferFunction = enum {
        pub const Custom: i32 = 0;
        pub const Linear: i32 = 1;
        pub const Gamma: i32 = 2;
        pub const SRgb: i32 = 3;
        pub const ProPhotoRgb: i32 = 4;
        pub const Bt2020: i32 = 5;
        pub const St2084: i32 = 6;
        pub const Hlg: i32 = 7;
    };

    pub const TransformModel = enum {
        pub const ThreeComponentMatrix: u8 = 0;
        pub const ElementListProcessing: u8 = 1;
    };

    pub const ColorModel = enum {
        pub const Undefined: u8 = 0;
        pub const Rgb: u8 = 1;
        pub const Gray: u8 = 2;
        pub const Cmyk: u8 = 3;
    };

};

File: binding/src/libqcolorspace.cpp
#include <QByteArray>
#include <QColorSpace>
#include <QColorTransform>
#include <QPointF>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QVariant>
#include <qcolorspace.h>
#include "libqcolorspace.h"
#include "libqcolorspace.hxx"

QColorSpace* QColorSpace_new() {
	 return new QColorSpace();
}

QColorSpace* QColorSpace_new2(int namedColorSpace) {
	 return new QColorSpace(static_cast<QColorSpace::NamedColorSpace>(namedColorSpace));
}

QColorSpace* QColorSpace_new3(QPointF* whitePoint, int transferFunction) {
	 return new QColorSpace(*whitePoint, static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new4(QPointF* whitePoint, const libqt_list transferFunctionTable) {
	 return new QColorSpace(*whitePoint, *transferFunctionTable);
}

QColorSpace* QColorSpace_new5(int primaries, int transferFunction) {
	 return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new6(int primaries, float gamma) {
	 return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), gamma);
}

QColorSpace* QColorSpace_new7(int primaries, const libqt_list transferFunctionTable) {
	 return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), *transferFunctionTable);
}

QColorSpace* QColorSpace_new8(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, int transferFunction) {
	 return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, static_cast<QColorSpace::TransferFunction>(transferFunction));
}

QColorSpace* QColorSpace_new9(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, const libqt_list transferFunctionTable) {
	 return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, *transferFunctionTable);
}

QColorSpace* QColorSpace_new10(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, const libqt_list redTransferFunctionTable, const libqt_list greenTransferFunctionTable, const libqt_list blueTransferFunctionTable) {
	 return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, *redTransferFunctionTable, *greenTransferFunctionTable, *blueTransferFunctionTable);
}

QColorSpace* QColorSpace_new11(const QColorSpace* colorSpace) {
	 return new QColorSpace(*colorSpace);
}

QColorSpace* QColorSpace_new12(QPointF* whitePoint, int transferFunction, float gamma) {
	 return new QColorSpace(*whitePoint, static_cast<QColorSpace::TransferFunction>(transferFunction), gamma);
}

QColorSpace* QColorSpace_new13(int primaries, int transferFunction, float gamma) {
	 return new QColorSpace(static_cast<QColorSpace::Primaries>(primaries), static_cast<QColorSpace::TransferFunction>(transferFunction), gamma);
}

QColorSpace* QColorSpace_new14(const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint, int transferFunction, float gamma) {
	 return new QColorSpace(*whitePoint, *redPoint, *greenPoint, *bluePoint, static_cast<QColorSpace::TransferFunction>(transferFunction), gamma);
}

void QColorSpace_OperatorAssign(QColorSpace* self, const QColorSpace* colorSpace) {
	self->operator=(*colorSpace);
}

void QColorSpace_Swap(QColorSpace* self, QColorSpace* colorSpace) {
	self->swap(*colorSpace);
}

int QColorSpace_Primaries(const QColorSpace* self) {
	return self->primaries();
}

int QColorSpace_TransferFunction(const QColorSpace* self) {
	return self->transferFunction();
}

float QColorSpace_Gamma(const QColorSpace* self) {
	return self->gamma();
}

libqt_string QColorSpace_Description(const QColorSpace* self) {
	QString _ret = self->description();
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QColorSpace_SetDescription(QColorSpace* self, const libqt_string description) {
	self->setDescription(QString::fromUtf8(description.data, description.len));
}

void QColorSpace_SetTransferFunction(QColorSpace* self, int transferFunction) {
	self->setTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction));
}

void QColorSpace_SetTransferFunction2(QColorSpace* self, const libqt_list transferFunctionTable) {
	self->setTransferFunction(*transferFunctionTable);
}

void QColorSpace_SetTransferFunctions(QColorSpace* self, const libqt_list redTransferFunctionTable, const libqt_list greenTransferFunctionTable, const libqt_list blueTransferFunctionTable) {
	self->setTransferFunctions(*redTransferFunctionTable, *greenTransferFunctionTable, *blueTransferFunctionTable);
}

QColorSpace* QColorSpace_WithTransferFunction(const QColorSpace* self, int transferFunction) {
	return new QColorSpace(self->withTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction)));
}

QColorSpace* QColorSpace_WithTransferFunction2(const QColorSpace* self, const libqt_list transferFunctionTable) {
	return new QColorSpace(self->withTransferFunction(*transferFunctionTable));
}

QColorSpace* QColorSpace_WithTransferFunctions(const QColorSpace* self, const libqt_list redTransferFunctionTable, const libqt_list greenTransferFunctionTable, const libqt_list blueTransferFunctionTable) {
	return new QColorSpace(self->withTransferFunctions(*redTransferFunctionTable, *greenTransferFunctionTable, *blueTransferFunctionTable));
}

void QColorSpace_SetPrimaries(QColorSpace* self, int primariesId) {
	self->setPrimaries(static_cast<QColorSpace::Primaries>(primariesId));
}

void QColorSpace_SetPrimaries2(QColorSpace* self, const QPointF* whitePoint, const QPointF* redPoint, const QPointF* greenPoint, const QPointF* bluePoint) {
	self->setPrimaries(*whitePoint, *redPoint, *greenPoint, *bluePoint);
}

void QColorSpace_SetWhitePoint(QColorSpace* self, QPointF* whitePoint) {
	self->setWhitePoint(*whitePoint);
}

QPointF* QColorSpace_WhitePoint(const QColorSpace* self) {
	return new QPointF(self->whitePoint());
}

uint8_t QColorSpace_TransformModel(const QColorSpace* self) {
	return self->transformModel();
}

int QColorSpace_ColorModel(const QColorSpace* self) {
	return self->colorModel();
}

void QColorSpace_Detach(QColorSpace* self) {
	self->detach();
}

bool QColorSpace_IsValid(const QColorSpace* self) {
	return self->isValid();
}

bool QColorSpace_IsValidTarget(const QColorSpace* self) {
	return self->isValidTarget();
}

QColorSpace* QColorSpace_FromIccProfile(const libqt_string iccProfile) {
	return new QColorSpace(QColorSpace::fromIccProfile(QByteArray(iccProfile.data, iccProfile.len)));
}

libqt_string QColorSpace_IccProfile(const QColorSpace* self) {
	QByteArray _qb = self->iccProfile();
	libqt_string _str;
	_str.len = _qb.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _qb.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

QColorTransform* QColorSpace_TransformationToColorSpace(const QColorSpace* self, const QColorSpace* colorspace) {
	return new QColorTransform(self->transformationToColorSpace(*colorspace));
}

QVariant* QColorSpace_OperatorQVariant(const QColorSpace* self) {
	return new QVariant(self->operator QVariant());
}

void QColorSpace_SetTransferFunction22(QColorSpace* self, int transferFunction, float gamma) {
	self->setTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction), gamma);
}

QColorSpace* QColorSpace_WithTransferFunction22(const QColorSpace* self, int transferFunction, float gamma) {
	return new QColorSpace(self->withTransferFunction(static_cast<QColorSpace::TransferFunction>(transferFunction), gamma));
}

void QColorSpace_Delete(QColorSpace* self) {
    delete self;
}

